runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1130 @@
+			..Default::default()
+		};
+
+		new_test_ext(genesis_config).execute_with(|| {
+			let id_a = ParaId::from(0u32);
+			let id_b = ParaId::from(1u32);
+
+			Paras::note_past_code(id_a, 10, 12, vec![1, 2, 3].into());
+			Paras::note_past_code(id_b, 20, 23, vec![4, 5, 6].into());
+
+			assert_eq!(<Paras as Store>::PastCodePruning::get(), vec![(id_a, 12), (id_b, 23)]);
+			assert_eq!(
+				Paras::past_code_meta(&id_a),
+				ParaPastCodeMeta {
+					upgrade_times: vec![upgrade_at(10, 12)],
+					last_pruned: None,
+				}
+			);
+			assert_eq!(
+				Paras::past_code_meta(&id_b),
+				ParaPastCodeMeta {
+					upgrade_times: vec![upgrade_at(20, 23)],
+					last_pruned: None,
+				}
+			);
+		});
+	}
+
+	#[test]
+	fn code_upgrade_applied_after_delay() {
+		let acceptance_period = 10;
+		let validation_upgrade_delay = 5;
+
+		let paras = vec![
+			(0u32.into(), ParaGenesisArgs {
+				parachain: true,
+				genesis_head: Default::default(),
+				validation_code: vec![1, 2, 3].into(),
+			}),
+		];
+
+		let genesis_config = MockGenesisConfig {
+			paras: GenesisConfig { paras, ..Default::default() },
+			configuration: crate::configuration::GenesisConfig {
+				config: HostConfiguration {
+					acceptance_period,
+					validation_upgrade_delay,
+					..Default::default()
+				},
+				..Default::default()
+			},
+			..Default::default()
+		};
+
+		new_test_ext(genesis_config).execute_with(|| {
+			let para_id = ParaId::from(0);
+			let new_code = ValidationCode(vec![4, 5, 6]);
+
+			run_to_block(2, None);
+			assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+
+			let applied_after = {
+				// this parablock is in the context of block 1.
+				let applied_after = 1 + validation_upgrade_delay;
+				Paras::schedule_code_upgrade(para_id, new_code.clone(), applied_after);
+				Paras::note_new_head(para_id, Default::default(), 1);
+
+				assert!(Paras::past_code_meta(&para_id).most_recent_change().is_none());
+				assert_eq!(<Paras as Store>::FutureCodeUpgrades::get(&para_id), Some(applied_after));
+				assert_eq!(<Paras as Store>::FutureCode::get(&para_id), new_code);
+				assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+
+				applied_after
+			};
+
+			run_to_block(applied_after, None);
+
+			// the candidate is in the context of the parent of `applied_after`,
+			// thus does not trigger the code upgrade.
+			{
+				Paras::note_new_head(para_id, Default::default(), applied_after - 1);
+
+				assert!(Paras::past_code_meta(&para_id).most_recent_change().is_none());
+				assert_eq!(<Paras as Store>::FutureCodeUpgrades::get(&para_id), Some(applied_after));
+				assert_eq!(<Paras as Store>::FutureCode::get(&para_id), new_code);
+				assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+			}
+
+			run_to_block(applied_after + 1, None);
+
+			// the candidate is in the context of `applied_after`, and triggers
+			// the upgrade.
+			{
+				Paras::note_new_head(para_id, Default::default(), applied_after);
```

Comment:
Why run to `applied_afted + 1` and notify about `applied_after`?

Reply:
This test and others like it was ported over from the old `parachains` module, where the equivalent function would always include a `CandidateReceipt` with the parent of the current chain head. 

Logically, since we are _within_ a block with number = `applied_after + 1`, there is no way to note a head which is executed in the context of `applied_after`, so this is more realistic of how the module should be used (although as I mentioned in another comment, it should really be applied_after + 2 for maximum realism).

But this test would work, I believe, even with the `execution_context` and current block number being the same, even though that would never happen in practice.